internal/app/cache/cache.go

```diff
@@ -1,19 +1,137 @@
 package cache
 
 import (
-	envoy_api_v2 "github.com/envoyproxy/go-control-plane/envoy/api/v2"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/golang/groupcache/lru"
+
+	v2 "github.com/envoyproxy/go-control-plane/envoy/api/v2"
 )
 
 type Cache interface {
-	// Exists returns true if the key exists in the cache, false otherwise.
-	Exists(key string) bool
-
 	// Fetch returns the cached response if it exists.
-	Fetch(key string) (*envoy_api_v2.DiscoveryResponse, error)
+	Fetch(key string) (*v2.DiscoveryResponse, error)
+
+	// SetResponse sets the cache response and returns the list of requests.
+	SetResponse(key string, resp v2.DiscoveryResponse) ([]*v2.DiscoveryRequest, error)
+
+	// AddRequest adds the request to the cache and returns whether a stream is open.
+	AddRequest(key string, req v2.DiscoveryRequest) (bool, error)
+}
+
+type cache struct {
+	cacheMu sync.RWMutex
+	cache   lru.Cache
+	ttl     time.Duration
+}
+
+type resource struct {
+	resp           *v2.DiscoveryResponse
+	requests       []*v2.DiscoveryRequest
+	streamOpen     bool
+	expirationTime time.Time
+}
 
-	// SetResponse sets the cache response and returns the list of open watches.
-	SetResponse(key string, resp envoy_api_v2.DiscoveryResponse) ([]*envoy_api_v2.DiscoveryRequest, error)
+// Callback function for each eviction. Receives the key and cache value when called.
+type onEvictFunc func(key lru.Key, value interface{})
+
+func NewCache(maxEntries int, onEvicted onEvictFunc, ttl time.Duration) (Cache, error) {
+	if ttl < 0 {
+		return nil, fmt.Errorf("ttl must be nonnegative but was set to %v", ttl)
+	}
+	return &cache{
+		cache: lru.Cache{
+			// Max number of cache entries before an item is evicted. Zero means no limit.
+			MaxEntries: maxEntries,
+			// OnEvict is called for each eviction and closes the stream if a key is removed
+			// (e.g. expiry due to TTL, too many entries).
```

Comment:
would remove `and closes the stream if a key is removed (e.g. expiry due to TTL, too many entries).` from here, since this is just the callback func.